File: python/phone-number/phone_number.py
import string


class PhoneNumber(object):
  def __init__(self, num):
    if any(c in num for c in string.ascii_letters):
      raise ValueError('letters not permitted')
    if any(c in num for c in set(string.punctuation) - set("()+-.")):
      raise ValueError('punctuations not permitted')
        

    # Ignore all non-digits.
    digits = ''.join([i for i in num if i.isdigit()])

    if len(digits) > 11:
      raise ValueError('must not be greater than 11 digits')
    # For an 11-digit with country code, drop the country code.
    if len(digits) == 11:
      if digits.startswith('1'):
        digits = digits[1:]
      else:
        raise ValueError('11 digits must start with 1')
    self.number = digits

    if len(self.number) != 10:
      raise ValueError('must not be fewer than 10 digits')

    for num, name in (('0', 'zero'), ('1', 'one')):
      for val, part in ((self.area_code, 'area'), (self.exchange, 'exchange')):
        if val.startswith(num):
          raise ValueError(f'{part} code cannot start with {name}')

  @property
  def area_code(self):
    return self.number[0:3]

  @property
  def exchange(self):
    return self.number[3:6]

  @property
  def subscriber(self):
    return self.number[6:10]

  def pretty(self):
    return '(%s)-%s-%s' % (self.number[0:3], self.number[3:6], self.number[6:10])


# vim:ts=2:sw=2:expandtab

File: go/phone-number/phone_number.go
// Package phonenumber cleans up phone numbers.
package phonenumber

import (
	"errors"
	"fmt"
)

// Number cleans and validates a phone number.
func Number(s string) (string, error) {
	t := make([]rune, 0, 12)
	for _, c := range s {
		if c < '0' || c > '9' {
			continue
		}
		t = append(t, c)
		if len(t) > 10 {
			break
		}
	}
	s = string(t)

	if len(s) == 11 {
		if s[0] == '1' {
			s = s[1:]
		} else {
			return "", errors.New("10 digit numbers must start with a '1'")
		}
	}
	if len(s) != 10 {
		return "", errors.New("number must be 9 or 10 digits")
	}
	if s[0] == '0' || s[0] == '1' {
		return "", errors.New("invalid area code")
	}
	if s[3] == '0' || s[3] == '1' {
		return "", errors.New("invalid exchange code")
	}
	return s, nil
}

// AreaCode returns a number's area code.
func AreaCode(s string) (string, error) {
	s, err := Number(s)
	if err != nil {
		return "", err
	}
	return s[0:3], nil
}

// Format formats a number all pretty-like.
func Format(s string) (string, error) {
	s, err := Number(s)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("(%s) %s-%s", s[0:3], s[3:6], s[6:10]), nil
}
